src/libos/src/vm/vm_manager.rs

```diff
@@ -440,31 +480,68 @@ impl VMManager {
                 let prefered_new_range =
                     VMRange::new_with_size(old_addr + old_size, new_size - old_size)?;
                 if self.is_free_range(&prefered_new_range) {
+                    // Don't need to move the old range
+                    let vm_initializer_for_new_range =
+                        if let Some((backed_file, offset)) = writeback_file {
+                            // Write back file should start from new offset
+                            writeback_file =
+                                Some((backed_file.clone(), offset + containing_vma.size()));
+                            VMInitializer::LoadFromFile {
+                                file: backed_file,
+                                offset: offset + containing_vma.size(), // file-backed mremap should start from the end of previous mmap/mremap file
+                            }
+                        } else {
+                            VMInitializer::FillZeros()
+                        };
                     let mmap_ops = VMMapOptionsBuilder::default()
                         .size(prefered_new_range.size())
                         .addr(VMMapAddr::Need(prefered_new_range.start()))
                         .perms(perms)
-                        .initializer(VMInitializer::FillZeros())
+                        .initializer(vm_initializer_for_new_range)
+                        .writeback_file(writeback_file)
                         .build()?;
                     (Some(mmap_ops), Some(old_addr))
                 } else {
+                    // Need to move old range to a new range and init the new range
+                    let vm_initializer_for_new_range =
+                        if let Some((backed_file, offset)) = &writeback_file {
+                            VMInitializer::CopyOldAndReadNew {
+                                old_range: old_range,
+                                file: backed_file.clone(),
+                                offset: offset + containing_vma.size(),
```

Comment:
Unlike other branches, the calculation of this offset in this branch is wrong. (Hint. There is a reason that `containing_vma` is NOT named `remapped_vma`. That is, `containing_vma` contains the range `[old_addr, old_addr + old_size)`.)

Reply:
After some digging, looks like we can't support file-backed mremap with sub range if don't support paging. I add a FIXME in the comment. Please take a look. Thank you for your revew.